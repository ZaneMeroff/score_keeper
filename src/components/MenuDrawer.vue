<template>

  <div>

    <q-header elevated :style="stateSettings.getDarkMode ? 'background-color: #000' : 'background-color: #FFF'">
      <q-toolbar>
        <q-btn flat :color="stateSettings.getDarkMode ? 'white' : 'black'" @click="drawer = !drawer" round dense icon="menu"></q-btn>
        <q-toolbar-title :style="stateSettings.getDarkMode ? 'color: #FFF' : 'color: #000'">Score Keeper</q-toolbar-title>
      </q-toolbar>
    </q-header>

    <q-drawer
      bordered
      overlay
      behavior="mobile"
      v-model="drawer"
      :dark="stateSettings.getDarkMode"
      :width="200"
    >
      <q-scroll-area class="fit">
        <q-list>

          <template v-for="(menuItem, index) in menuList">
            <q-item :key="index" @click="menuItem.onClick" clickable v-ripple>
              <q-item-section avatar>
                <q-icon :name="menuItem.icon"></q-icon>
              </q-item-section>
              <q-item-section>{{ menuItem.label }}</q-item-section>
            </q-item>
            <q-separator :key="'sep' + index" v-if="menuItem.separator"></q-separator>
          </template>

        </q-list>
      </q-scroll-area>
    </q-drawer>

    <!-- score limits modal -->
    <score-limits 
      :showModal="stateModals.getShowScoreLimitsModal"
    />

    <!-- add players modal -->
    <add-players
      :showModal="stateModals.getShowAddPlayersModal"
    />

    <!-- reset score confirm modal -->
    <confirm
      :showModal="stateModals.getShowResetScoreModal"
      :text="'Reset all scores to zero?'"
      :onYes="() => {
        statePlayers.action_zeroScores()
        stateModals.action_resetScoreModalVisibility(false)
      }"
      :onNo="() => stateModals.action_resetScoreModalVisibility(false)"
    />

    <!-- delete players confirm modal -->
    <confirm
      :showModal="stateModals.getShowDeletePlayersModal"
      :text="'Delete all players?'"
      :onYes="() => {
        statePlayers.action_deleteAllPlayers()
        stateModals.action_deletePlayersModalVisibility(false)
      }"
      :onNo="() => stateModals.action_deletePlayersModalVisibility(false)"
    />

    <!-- clear data confirm modal -->
    <confirm
      :showModal="stateModals.getShowClearDataModal"
      :text="'Clear all app data?'"
      :onYes="() => handleClearData()"
      :onNo="() => stateModals.action_clearDataModalVisibility(false)"
    />

  </div>

</template>

<script lang="ts">

// Decorators
import { Component, Vue } from 'vue-property-decorator'

// Vuex
import { stateModals, statePlayers, stateSettings } from '@/store/index'

// Components
import AddPlayers from './modals/AddPlayers.vue'
import Confirm from './modals/Confirm.vue'
import ScoreLimits from './modals/ScoreLimits.vue'

// Utils
import localforage from 'localforage'

@Component({
  components: {
    'add-players': AddPlayers,
    'confirm': Confirm,
    'score-limits': ScoreLimits
  }
})
export default class MenuDrawer extends Vue {

  stateModals = stateModals
  statePlayers = statePlayers
  stateSettings = stateSettings
  

  drawer: boolean = false

  async handleClearData() {
    await localforage.clear()
      .then(() => stateModals.action_clearDataModalVisibility(false))
      .catch(err => window.alert(`Whoops, an error occured: ${err}`))
  }

  // create typed for menuList item !!!
  menuList = [
    {
      icon: 'group_add',
      label: 'Add Players',
      onClick: () => {
        this.stateModals.action_addPlayersModalVisibility(true)
      },
      separator: true
    },
    {
      icon: 'sports_score',
      label: 'Score Limits',
      onClick: () => {
        this.stateModals.action_scoreLimitModalVisibility(true)
      },
      separator: true
    },
    {
      icon: 'restart_alt',
      label: 'Zero Scores',
      onClick: () => {
        this.stateModals.action_resetScoreModalVisibility(true)
      },
      separator: true
    },
    {
      icon: 'delete',
      label: 'Delete Players',
      onClick: () => {
        this.stateModals.action_deletePlayersModalVisibility(true)
      },
      separator: true
    },
    {
      icon: 'dark_mode',
      label: 'Toggle Theme',
      onClick: () => {
        this.stateSettings.action_setDarkMode()
      },
      separator: true
    },
    {
      icon: 'phonelink_erase',
      label: 'Clear Data',
      onClick: () => {
        this.stateModals.action_clearDataModalVisibility(true)
      },
      separator: true
    },
  ]
}
</script>