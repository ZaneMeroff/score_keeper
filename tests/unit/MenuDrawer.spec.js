import { createLocalVue, shallowMount } from '@vue/test-utils'
import MenuDrawer from '@/components/MenuDrawer.vue'
import Quasar, * as All from 'quasar'
import Vuex from 'vuex'
import localforage from 'localforage'

// --------- Vuex & Quasar setup -------------

const localVue = createLocalVue()
localVue.use(Vuex, Quasar, {components: All, directives: All, plugins: All})

// -------------------------------------------

describe('MenuDrawer', () => {

  describe('snapshots', () => {

    it('should render with default html', () => {
      const component = shallowMount(MenuDrawer, { localVue })
      
      expect(component.element).toMatchSnapshot()
    })
  })

  describe('data', () => {

    describe('drawer', () => {

      it('should have a default value of false', () => {
        const component = shallowMount(MenuDrawer, { localVue })
        
        expect(component.vm.$data.drawer).toEqual(false)
      })

      it('should accept a boolean', () => {
        const component = shallowMount(MenuDrawer, { localVue })
        const newValue = true

        component.setData({ drawer: newValue })

        expect(component.vm.$data.drawer).toEqual(newValue)
      })
    })

    describe('menuList', () => {

      it('should have a 6 menuItems', () => {
        const component = shallowMount(MenuDrawer, { localVue })
        
        expect(component.vm.$data.menuList.length).toEqual(6)
      })
    })
  })

  describe('methods', () => {

    describe('handleClearData', () => {

      it('should call localforage.clear()', () => {
        const component = shallowMount(MenuDrawer, { localVue })
        const spy = jest.spyOn(localforage, 'clear').mockImplementation(() => {})

        component.vm.handleClearData()
        
        expect(spy).toHaveBeenCalledTimes(1)
      })
    })
  })
})
