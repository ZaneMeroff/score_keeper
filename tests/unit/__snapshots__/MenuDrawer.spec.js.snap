// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuDrawer snapshots should render with default html 1`] = `
<div>
  <q-header-stub
    elevated="true"
    heighthint="50"
    revealoffset="250"
    style="background-color: rgb(0, 0, 0);"
    value="true"
  >
    <q-toolbar-stub>
      <q-btn-stub
        align="center"
        dense="true"
        flat="true"
        icon="menu"
        ripple="true"
        round="true"
        style="color: rgb(255, 255, 255);"
      />
       
      <q-toolbar-title-stub
        style="color: rgb(255, 255, 255);"
      >
        Score Keeper
      </q-toolbar-title-stub>
    </q-toolbar-stub>
  </q-header-stub>
   
  <q-drawer-stub
    behavior="mobile"
    bordered="true"
    breakpoint="1023"
    dark="true"
    miniwidth="57"
    overlay="true"
    side="left"
    width="200"
  >
    <q-scroll-area-stub
      class="fit"
      delay="1000"
    >
      <q-list-stub>
        <q-item-stub
          clickable="true"
          tag="div"
        >
          <q-item-section-stub
            avatar="true"
          >
            <q-icon-stub
              name="group_add"
              tag="i"
            />
          </q-item-section-stub>
           
          <q-item-section-stub>
            Add Players
          </q-item-section-stub>
        </q-item-stub>
         
        <q-separator-stub />
        <q-item-stub
          clickable="true"
          tag="div"
        >
          <q-item-section-stub
            avatar="true"
          >
            <q-icon-stub
              name="sports_score"
              tag="i"
            />
          </q-item-section-stub>
           
          <q-item-section-stub>
            Score Limits
          </q-item-section-stub>
        </q-item-stub>
         
        <q-separator-stub />
        <q-item-stub
          clickable="true"
          tag="div"
        >
          <q-item-section-stub
            avatar="true"
          >
            <q-icon-stub
              name="restart_alt"
              tag="i"
            />
          </q-item-section-stub>
           
          <q-item-section-stub>
            Zero Scores
          </q-item-section-stub>
        </q-item-stub>
         
        <q-separator-stub />
        <q-item-stub
          clickable="true"
          tag="div"
        >
          <q-item-section-stub
            avatar="true"
          >
            <q-icon-stub
              name="delete"
              tag="i"
            />
          </q-item-section-stub>
           
          <q-item-section-stub>
            Delete Players
          </q-item-section-stub>
        </q-item-stub>
         
        <q-separator-stub />
        <q-item-stub
          clickable="true"
          tag="div"
        >
          <q-item-section-stub
            avatar="true"
          >
            <q-icon-stub
              name="dark_mode"
              tag="i"
            />
          </q-item-section-stub>
           
          <q-item-section-stub>
            Toggle Colors
          </q-item-section-stub>
        </q-item-stub>
         
        <q-separator-stub />
        <q-item-stub
          clickable="true"
          tag="div"
        >
          <q-item-section-stub
            avatar="true"
          >
            <q-icon-stub
              name="phonelink_erase"
              tag="i"
            />
          </q-item-section-stub>
           
          <q-item-section-stub>
            Clear Data
          </q-item-section-stub>
        </q-item-stub>
         
        <q-separator-stub />
      </q-list-stub>
    </q-scroll-area-stub>
  </q-drawer-stub>
   
  <score-limits-stub />
   
  <add-players-stub />
   
  <confirm-stub
    onno="[Function]"
    onyes="[Function]"
    text="Reset all scores to zero?"
  />
   
  <confirm-stub
    onno="[Function]"
    onyes="[Function]"
    text="Delete all players?"
  />
   
  <confirm-stub
    onno="[Function]"
    onyes="[Function]"
    text="Clear all app data?"
  />
</div>
`;
